e2e/investigation.md

```diff
@@ -0,0 +1,96 @@
+    There is no cleanup process implemented for individual e2e tests that would restore the dev-env to a state
+    from where other e2e tests could be started, this is partly the case because the restoration process is highly
+    dependent on the e2e test that was executed as some tests affect the configuration of the whole dev-env.
+
+It has become clear that there are multiple ways to achieve the parallel test execution
+
+## Possible solutions
+
+**Option 1 (pool of dev-envs)**:
+
+    This solution requires the bootstrapping of separate dev-env instances
+    (both management cluster and target cluster) for individual e2e test cases.
+    (CON: it scales poorly and it would cause problems with locally running tests on a developer's machine)
+
+**Option 2 (solve it in CI)**:
+
+    This approach would require minimal modification to the CAPM3 Makefile and
+    additional some additional logic to the test code to implement the ability to selectively execute e2e test
+    specs (similarly how it is in dev-env integration tests). Option 3 also includes the individually executable tests.
+
+    This approach would basically be implemented by running multiple e2e test jobs in CI that would only test individual e2e
+    test cases. (CON: this would still require multiple hosts in CityCloud and it would be still slow on local dev-envs).
+
+**Option3 (GO approach)**:
+
+    Ginkgo test specs can be executed in a parallel fashion as described in https://onsi.github.io/ginkgo/#parallel-specs .
+
+    Each individual test case should have it's own cleanup process that reverts the modifications
+    done by the test execution.
+
+    The cleanup process should aim to avoid re-installation of the
+    dev-env. In some cases like pivoting the cleanup process could be implemented as an other
+    test spec/case e.g "re-pivoting".
+
+    Test specs that are executed in parallel mode should be grouped separately from the ones executed sequentially as
+    not all e2e test cases work well together. Alternatively tests specs could be synced but
+    that could add considerable cognitive burden when there are a lot of tests.
+
+    All the test cases should be able run individually e.g "make test-e2e pivoting update" etc...
+    this could be useful in the future when there will be more test cases.
+
+**NOTE**:
+
+	There is the possibility to combine option 3 and 1 such as in case the modifications is done as part of
+	option 3 would not yield the needed time savings then the test process could be extended with the use of
+	multiple dev-envs and tests could be executed across multiple dev-envs if needed. With this approach
+	the test execution would be still handled by ginkgo parallel test approach and it would require a
+	"big" virtual machine as a build host. This approach also creates a lot of technical depth and increases
+	the complexity of the tests. CAPO does something similar to this.
```

Comment:
What about running the tests in different Kubernetes namespaces, instead of in different dev envs? That's what CAPO does. Each CAPI cluster is in a different namespace of the management cluster, and each Gingko `Describe` block runs in parallel (2 at a time) spinning up a machine + cluster from the same source k8s. See: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/5e92c921936140cdfaa870b10bc7a1d6f2c0b91e/test/e2e/shared/common.go#L41. 

Here you seem to be suggesting different that the tests run in different metal3-dev-envs, but did you consider using a single dev env with multiple namespaces in the management cluster? (A large machine would still be required for the pool of VMs.)

Reply:
Yes I have considered it, there are always certain tests that can be run in a parallel manner next to each other, but there are disruptive ones like pivot, repivot, upgrade that I think would cause instability for other tests. 

But running test parallell in different namespace is perfectly fine and this is  the reason why I mentioned this:

" Test specs that are executed in parallel mode should be grouped separately from the ones executed sequentially as not all e2e test cases work well together. Alternatively tests specs could be synced but that could add considerable cognitive burden when there are a lot of tests."
    
  But I think you are right It could be explained better. I will rewrite it a bit with this in mind. I think I can highlight that the parallel execution would be in separate namespaces on the same dev-env.

Reply:
I think upgrade would probably be OK, since the target nodes will live in their own namespace, but pivot and repivot would definitely require some thought.